### DISCUSSION TITLE
Entities with wildcard schema and migration

### DISCUSSION THREAD

**Discussion by pphunor72** [mikro-orm/mikro-orm#4929] [General]
Currently there is no support to generate migration file for entites with wildcard schema. As far as I see generating a migration file does not need active database connection, it works based on snapshot file pretty well.

If no schema specified the migration file will contain SQL statements without any direct schema reference. Currently I am using this approach instead of wildcard schema, but it leads to runtime problems. Usually these entities are referred with schema specified in options (FindOptions for example), but sometimes not (default schema is used for example at population with select-in strategy).

What if entities with wildcard schema behave like entities without any schema specification during generation of migration file? It can work based on snapshot and the resulting SQL statements will not contain any schema reference. Before applying such a migration file I can set the proper default schema, and the migration will work as expected.

**Comment by B4nan** [comment]
> Currently there is no support to generate migration file for entites with wildcard schema. As far as I see generating a migration file does not need active database connection, it works based on snapshot file pretty well.

Snapshots can be disabled (or manually removed), but yes, if it's there, it's preferred.

> What if entities with wildcard schema behave like entities without any schema specification during generation of migration file? It can work based on snapshot and the resulting SQL statements will not contain any schema reference. Before applying such a migration file I can set the proper default schema, and the migration will work as expected.

You can't use migrations for those, migrations are static, wildcard schemas are dynamic - you can add them at any point and the existing migrations cannot take this into account (and this can vary from environemt, I doubt you would have the same schemas in dev env and in production, so this would completely break the migrations workflow, where you test them first locally, then dev, then prod - the very same migration files). There are several older issues around this topic, e.g. #3319. I don't want a half-baked solution, and so far nobody suggested anything thorough. IMO you should be using the schema generator directly for those.

**Reply by pphunor72** [reply]
I don't really understand this "dynamic" nature of wildcard schemas. In code first approach which I use, I have a set of entities, most of them with wildcard schema. Whenever I change them I generate two migration scripts: one for wildcard schemas and one for explicitly set schemas. Script of wildcard schema is executed several times, once for every tenant schema.
Working with snapshot means that the database structure is stored under source control together with entities, and this is much suitable for me than a template database the migration script should be generated against (use of SchemaGenerator).

**Reply by B4nan** [reply]
> once for every tenant schema.

The tenants are what is dynamic, you can add a new one over time, and often this happens dynamically (not on deploy where you would run migrations).

**Reply by pphunor72** [reply]
OK, but our application adds new schemas (representing new tenants) and also ensures that the migration scripts are executed. I have never heard of such a scenario where the application has to handle schemas that suddenly appear in the database independently of it.
